### PR TITLE
[DRAFT] Add filelogexporter.

### PR DESCRIPTION
**Pull request by hypnoce**
This exporter plays the same role as the file log receiver but append-writes to files.

**Description:** <Describe what has changed.>
This PR adds a new file log exporter. It differs from the fileexporter in that it can write to different files based on resources and/or log attributes. It is the equivalent exporter of the file log receiver.

**Testing:** Happy path and failures were tested.

**Documentation:** README with configuration description and exemple 
Naming strategy is the following 
image

### CODE REVIEW

File: exporter/filelogexporter/config.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	"errors"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+// Config defines configuration for file exporter.
+type Config struct {
+	config.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+
+	// Path to write to. Path is relative to current directory.
```
Comment (jcchavezs) [inline exporter/filelogexporter/config.go:27]:
```suggestion
	// Path to write to. Path is relative to current working directory.
```

File: exporter/filelogexporter/config.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+// Validate checks if the exporter configuration is valid
+func (cfg *Config) Validate() error {
+	if cfg.Path == "" {
+		return errors.New("path must be non-empty")
+	}
+	if cfg.FilenameAttributeKey == "" {
+		return errors.New("file_name_attr must be non-empty")
+	}
+	if cfg.MaxOpenFiles < 0 {
```
Comment (jcchavezs) [inline exporter/filelogexporter/config.go:47]: Is `0` valid?

File: exporter/filelogexporter/exporter.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	path                 string
+	basedirAttributeKey  string
+	filenameAttributeKey string
+	maxOpenFiles         int
+	files                *lru.Cache
+	mutex                sync.Mutex
+	logger               *zap.Logger
+}
+
+func (e *fileLogExporter) Capabilities() consumer.Capabilities {
```
Comment (jcchavezs) [inline exporter/filelogexporter/exporter.go:43]:
```suggestion
func (fileLogExporter) Capabilities() consumer.Capabilities {
```

File: exporter/filelogexporter/exporter.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+		for j := 0; j < ills.Len(); j++ {
+			libraryLog := ills.At(j)
+			lls := libraryLog.Logs()
+			for k := 0; k < lls.Len(); k++ {
+				log := lls.At(k)
+				logBase := resourceBase
+				if baseDir, exists := log.Attributes().Get(e.basedirAttributeKey); exists {
+					logBase = baseDir.AsString()
+				}
+				if fileName, exists := log.Attributes().Get(e.filenameAttributeKey); exists {
```
Comment (jcchavezs) [inline exporter/filelogexporter/exporter.go:67]: nit: this body is a bit long, can we do early return inverting the condition?

File: exporter/filelogexporter/config.go (hunk 2 above)
Comment (hypnoce) [reply exporter/filelogexporter/config.go:47]: 0 means unlimited. I can add this to the README.

File: exporter/filelogexporter/exporter_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/testdata"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+)
+
```
Comment (jcchavezs) [inline exporter/filelogexporter/exporter_test.go:32]: worth to add a test for when the log file can't be created?

Comment (hypnoce) [reply exporter/filelogexporter/exporter_test.go:32]: Done

File: exporter/filelogexporter/exporter.go (hunk 4 above)
Comment (hypnoce) [reply exporter/filelogexporter/exporter.go:67]: Not sure how it will make the body smaller. I can extract it in another method ?

Comment (jcchavezs) [reply exporter/filelogexporter/exporter.go:67]: I am sorry I did not explain myself correctly. I meant that the body is long and identation makes it harder to read (due to max line length so using the early return will reduce the identation.

Comment (hypnoce) [reply exporter/filelogexporter/exporter.go]: Done.

**Comment by bogdandrutu** [comment]
@djaglowski @tigrannajaryan we need to be careful with "file" receiver/exporter and when it comes to logs vs other sources. There are multiple requests see https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/4997

**Comment by hypnoce** [comment]
> @djaglowski @tigrannajaryan we need to be careful with "file" receiver/exporter and when it comes to logs vs other sources. There are multiple requests see #4997

I think we should consider the followings in separate issues : 

- File location where a log record is written (scope of this PR)
- File handling 
  - Create/open options (mod, truncate...)
  - Time and/or size constraints
  - Strategy when breaking constraints (rotation, discard, start from top...)
- File encoding
  - type of telemetry data (log, metric, trace)
  - encoding (json, binary, text, raw body...)

WDYT ?

**Comment by tigrannajaryan** [comment]
@hypnoce I would advise to convert this PR to draft, then open an issue that describes your needs. This needs to be discussed and we need to agree on a design and on a file format before moving forward.

Before we implement a filelog exporter first we need to decide on the format of the file. This is an open issue in the spec currently: https://github.com/open-telemetry/opentelemetry-specification/issues/1443

The exporter in this PR appears to write the Body of the log record, followed by a new line (from what I see). It seems to ignore all other attributes of the log record and of the Resource and also does not seem to do any escaping of values so it is not useful as a machine readable format (I may be wrong, I only quickly glanced at the PR). 

It is not clear why this is _the_ filelog exporter of this project. It is probably a filelog exporter that is useful for you. If you need it urgently for your own use it is best to implement the component in your own git repository and create a custom build of the Collector that includes this component.

For  the filelog exporter that is useful for the community we need to spend more time understanding and designing how this should work.

**Comment by hypnoce** [comment]
@tigrannajaryan I will open an issue gathering requirements.
> we need to decide on the format of the file.

We can decouple the file format as it can be configurable, and start with existing formats (json, binary body...), being machine or human readable.

> It is probably a filelog exporter that is useful for you.

This filelog exporter is the symmetric of the existing filelog receiver that is able to read any format. It was created with this receiver in mind thus having the same name (filelog) and respecting 'source' file encoding.

**Pull request closed without merging**